app/models/notification.rb

```diff
@@ -13,6 +13,7 @@
 #  from_account_id :bigint(8)        not null
 #  type            :string
 #  filtered        :boolean          default(FALSE), not null
+#  group_key       :string
```

Comment:
I wonder about the tradeoffs on the approach here (adding `group_key` string to Notification, the AMS Model, the new NotificationGroup serializer) vs any of:

- Something where the "grouped stats about activity on some object" concept was closer to a first-class object of its own and could be queried/paginated directly instead of going through notifications in that query. Might have similar callback needs to the "Stat" stuff on Status/Account with that approach.
- A materialized view that could have a more direct AR class wrapping up the data.

Do we need to be able to forever query the historical time windows of this data? Like, is there a use case for querying "what did this specific hour three months ago look like?" or is it more real-time/ephemeral and fine if we only have some recent window of data?

Reply:
> Something where the "grouped stats about activity on some object" concept was closer to a first-class object of its own and could be queried/paginated directly instead of going through notifications in that query. Might have similar callback needs to the "Stat" stuff on Status/Account with that approach.

Yeah, that is something I have been thinking about as well, but I'm afraid this would involve a lot of rather complex book-keeping open to concurrency issues…

> A materialized view that could have a more direct AR class wrapping up the data.

I am not too sure how that would look like, and materialized views, with `Scenic` in particular, have proved to be a massive pain in handling migration paths.

> Do we need to be able to forever query the historical time windows of this data? Like, is there a use case for querying "what did this specific hour three months ago look like?" or is it more real-time/ephemeral and fine if we only have some recent window of data?

Do you mean outright losing old notifications, or losing the way they are grouped?

Reply:
> Do you mean outright losing old notifications, or losing the way they are grouped?

More the latter. Like from the api perspective, can you keep paginating/querying infinitely far back (or I guess, up to the introduction of this feature) and expect to always have results? This is sort of in the context of the DB view approach, and whether you could have it built for a moving window time period (last X days?) which might keep the view generation more modest, easier to keep current, etc.

Reply:
I'm honestly not sure. I think if grouping is important to keep things readable, it remains important for the past as well. Possibly we could allow coarser grouping for older entries, but I don't think it would make things easier.